multipart-config & getParameter clarification

I apologize if this issue has been brought up already, but I'm wondering if there are any plans to address this in a future spec. 

Scenario: There is a `multipart-config` servlet with a `max-file-size` specified, and a file, which is over the limit,  is uploaded to this servlet. The servlet then first calls `getParameter` when processing the request. 

During parsing of the parameters, the servlet tries to parse the multi-part form data for any parameters but encounters an issue with the file being too large. Some exception is thrown, and, since parsing was incomplete, `getParameter` returns null (even if it was for a valid parameter).


My question is how should the servlet implementation handle this scenario?

- Should null continue to be returned (current behavior)?
- Should an `IllegalStateException` or some other exception be thrown to indicate something went wrong? 
- Some other option (i.e `getParts` should be checked first as explained in ryanchapin's post below) ? 

For context, this was discovered through a JSF (MyFaces) upload application.  MyFaces checks postback requests by looking for a view-state parameter but the uploaded file is over the max-file-size.  Hence, `getParameter` is null.  JSF thinks this is a new request.  The getParameter is called by the implementation (so the application hasn't had a chance to call getPart/getParts yet)

If I'm missing something here or there's an existing workaround, please let me know. Otherwise, I'm hoping a future specification may address this scenario. 

I found a old blog post here, so it seems like this issue has exist for a while: https://www.ryanchapin.com/invocation-of-request-getparts-in-a-servlet-3-0-dopost-method-will-not-throw-illegalstateexception/

Thanks! 

References: 

1) https://javaee.github.io/javaee-spec/javadocs/javax/servlet/ServletRequest.html#getParameter-java.lang.String-

2) https://javaee.github.io/javaee-spec/javadocs/javax/servlet/annotation/MultipartConfig.html 

3) 3.2 File upload  - Servlet 4.0 Spec

Invalid %xx decoding falls into 2 categories.

1. Bad UTF-8 Encodings (incomplete, not a valid sequence, not a glyph, overlong encoding, etc).  Basically the input is valid `%` [hex] [hex], but not pointing to something that can be encoded in Unicode (eg: `/foo/%EF%28`)
2. Something that isn't valid percent encoding (eg: `/foo/%xx` or `/bar/100%`)

For the first category, the unicode replacement character can be used. (like what is done with `java.nio.charset.CharsetDecoder`, `java.net.URLDecoder`, and `java.net.URI` classes in the JVM).

```
jshell> URI.create("/foo/%EF%28").getPath()
$1 ==> "/foo/�("
jshell> URLDecoder.decode("/foo/%EF%28")
$3 ==> "/foo/�("
```

For the second category, it's a fundamentally bad input, and should just be rejected.
```
jshell> URI.create("/foo/%xx").getPath()
|  Exception java.lang.IllegalArgumentException: Malformed escape pair at index 5: /foo/%xx
|        at URI.create (URI.java:906)
|        at (#2:1)
|  Caused by: java.net.URISyntaxException: Malformed escape pair at index 5: /foo/%xx
|        at URI$Parser.fail (URI.java:2974)
|        at URI$Parser.scanEscape (URI.java:3102)
|        at URI$Parser.scan (URI.java:3125)
|        at URI$Parser.checkChars (URI.java:3143)
|        at URI$Parser.parseHierarchical (URI.java:3227)
|        at URI$Parser.parse (URI.java:3186)
|        at URI.<init> (URI.java:623)
|        at URI.create (URI.java:904)
|        ...
jshell> URLDecoder.decode("/foo/%xx")
|  Exception java.lang.IllegalArgumentException: URLDecoder: Illegal hex characters in escape (%) pattern - Error at index 0 in: "xx"
|        at URLDecoder.decode (URLDecoder.java:237)
|        at URLDecoder.decode (URLDecoder.java:147)
|        at URLDecoder.decode (URLDecoder.java:109)
|        at (#4:1)
```

--- a/api/src/main/java/jakarta/servlet/ServletRequest.java
+++ b/api/src/main/java/jakarta/servlet/ServletRequest.java
@@ -164,9 +164,22 @@ public interface ServletRequest {
      * If the parameter data was sent in the request body, such as occurs with an HTTP POST request, then reading the body
      * directly via {@link #getInputStream} or {@link #getReader} can interfere with the execution of this method.
      *
+     * <p>
+     * If not already parsed, calling this method will trigger the parsing of the query string, POSTed form data where the
+     * request body has content type is <code>application/x-www-form-urlencoded</code> and POSTed form data where the
+     * request body has content-type <code>multipart/form-data</code> and the Servlet is configured with a
+     * {@link jakarta.servlet.annotation.MultipartConfig} annotation or a <code>multipart-config</code> element in the
+     * deployment descriptor.
+     *
      * @param name a <code>String</code> specifying the name of the parameter
      *
      * @return a <code>String</code> representing the single value of the parameter
+     *
+     * @throws IllegalStateException if parameter parsing is triggered and a problem is encountered parsing the parameters
+     * including, but not limited to: invalid percent (%nn) encoding; invalid byte sequence for the specified character set;
+     * I/O errors reading the request body; and triggering a container defined limit related to parameter parsing.
+     * Containers may provide container specific options to handle some or all of these errors in an alternative manner that
+     * may include not throwing an exception.
      *
      * @see #getParameterValues
      */
@@ -177,8 +190,21 @@ public interface ServletRequest {
      * Returns an <code>Enumeration</code> of <code>String</code> objects containing the names of the parameters contained
      * in this request. If the request has no parameters, the method returns an empty <code>Enumeration</code>.
      *
+     * <p>
+     * If not already parsed, calling this method will trigger the parsing of the query string, POSTed form data where the
+     * request body has content type is <code>application/x-www-form-urlencoded</code> and POSTed form data where the
+     * request body has content-type <code>multipart/form-data</code> and the Servlet is configured with a
+     * {@link jakarta.servlet.annotation.MultipartConfig} annotation or a <code>multipart-config</code> element in the
+     * deployment descriptor.
+     *
      * @return an <code>Enumeration</code> of <code>String</code> objects, each <code>String</code> containing the name of a
      * request parameter; or an empty <code>Enumeration</code> if the request has no parameters
+     *
+     * @throws IllegalStateException if parameter parsing is triggered and a problem is encountered parsing the parameters
+     * including, but not limited to: invalid percent (%nn) encoding; invalid byte sequence for the specified character set;
+     * I/O errors reading the request body; and triggering a container defined limit related to parameter parsing.
+     * Containers may provide container specific options to handle some or all of these errors in an alternative manner that
+     * may include not throwing an exception.
      */
     public Enumeration<String> getParameterNames();
 
@@ -189,9 +215,22 @@ public interface ServletRequest {
      * <p>
      * If the parameter has a single value, the array has a length of 1.
      *
+     * <p>
+     * If not already parsed, calling this method will trigger the parsing of the query string, POSTed form data where the
+     * request body has content type is <code>application/x-www-form-urlencoded</code> and POSTed form data where the
+     * request body has content-type <code>multipart/form-data</code> and the Servlet is configured with a
+     * {@link jakarta.servlet.annotation.MultipartConfig} annotation or a <code>multipart-config</code> element in the
+     * deployment descriptor.
+     *
      * @param name a <code>String</code> containing the name of the parameter whose value is requested
      *
      * @return an array of <code>String</code> objects containing the parameter's values
+     *
+     * @throws IllegalStateException if parameter parsing is triggered and a problem is encountered parsing the parameters
+     * including, but not limited to: invalid percent (%nn) encoding; invalid byte sequence for the specified character set;
+     * I/O errors reading the request body; and triggering a container defined limit related to parameter parsing.
+     * Containers may provide container specific options to handle some or all of these errors in an alternative manner that
+     * may include not throwing an exception.
      *
      * @see #getParameter
      */
@@ -204,8 +243,21 @@ public interface ServletRequest {
      * Request parameters are extra information sent with the request. For HTTP servlets, parameters are contained in the
      * query string or posted form data.
      *
+     * <p>
+     * If not already parsed, calling this method will trigger the parsing of the query string, POSTed form data where the
+     * request body has content type is <code>application/x-www-form-urlencoded</code> and POSTed form data where the
+     * request body has content-type <code>multipart/form-data</code> and the Servlet is configured with a
+     * {@link jakarta.servlet.annotation.MultipartConfig} annotation or a <code>multipart-config</code> element in the
+     * deployment descriptor.
+     *
      * @return an immutable java.util.Map containing parameter names as keys and parameter values as map values. The keys in
      * the parameter map are of type String. The values in the parameter map are of type String array.
+     *
+     * @throws IllegalStateException if parameter parsing is triggered and a problem is encountered parsing the parameters
+     * including, but not limited to: invalid percent (%nn) encoding; invalid byte sequence for the specified character set;
+     * I/O errors reading the request body; and triggering a container defined limit related to parameter parsing.
+     * Containers may provide container specific options to handle some or all of these errors in an alternative manner that
+     * may include not throwing an exception.
      */
     public Map<String, String[]> getParameterMap();
 

--- a/spec/src/main/asciidoc/servlet-spec-body.adoc
+++ b/spec/src/main/asciidoc/servlet-spec-body.adoc
@@ -8614,6 +8614,13 @@ Add new constants to `HttpServletResponse` for the HTTP response codes 308, 421,
 link:https://github.com/eclipse-ee4j/servlet-api/issues/415[Issue 415]::
 Add overloaded `setCharacterEncoding()` methods that support `Charset`.
 
+link:https://github.com/eclipse-ee4j/servlet-api/issues/431[Issue 431]::
+`ServletRequest.getParameter()` and the other parameter methods are now
+documented to throw a runtime exception if the call triggers parsing of the
+parameters and an error is encountered during that parsing. Containers may
+provide container specific options to handle some or all of such errors in
+an alternative manner that may include not throwing the runtime exception.  
+
 link:https://github.com/eclipse-ee4j/servlet-api/issues/443[Issue 443]::
 Remove references to the SecurityManager and associated APIs.
 

